File: pset6/credit/credit.py
from cs50 import get_int
# Function that prompts user for credit number


def get_number():

    number = get_int("Number: ")
    return number

# Function that returns the first 2 digits of the card


def get_digits(number):
    while number >= 100:
        number = number / 10
    return number

# Function that returns the first digit of a credit card


def get_digit(number):
    while number >= 10:
        number = int(number / 10)
    return number

# Function that determines if credit card number is valid


def credit():

    # Get the credit card number
    number = get_number()
    checkSum = 0
    count = 0

    # Make a copy of the credit card number
    cardCopy = number

    # Store first two digits of card for type checking
    digits = int(get_digits(cardCopy))

    # Looping through the digits and adding every second digit to the checkSum afterwards
    while number > 0:
        count += 1

        # If even digit multiply the remainder by 2 and add the product digits to checkSum
        if count % 2 == 0:
            r = 2 * int(number % 10)

            # If r has more than one digit then add all digits to checkSum
            if len(str(r)) > 1:
                while r > 0:
                    checkSum += int(r % 10)
                    r = int(r / 10)
            else:
                checkSum += r

        # If odd digit then add remainder to checkSum
        else:
            checkSum += int(number % 10)

        number = int(number / 10)

    # Verfiy if the card is an AMEX
    if count == 15 and (digits == 34 or digits == 37):
        card = "AMEX"

    # Verify if card is of type VISA
    elif get_digit(cardCopy) == 4 and (count == 13 or count == 16):
        card = "VISA"

    # Verify if card is of type MASTERCARD
    elif count == 16 and (digits == 51 or digits == 52 or digits == 53 or digits == 54 or digits == 55):
        card = "MASTERCARD"

    # Otherwise the card number is invalid
    else:
        card = "INVALID"

    # Printing the card type
    if checkSum % 10 != 0:
        print("INVALID")

    else:
        print(card)


credit()

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

//Calling methods defined after the main method
long get_credit_number(void);
int get_length(long n);
void get_card_type(long n);
int get_digit_sum(long num);
int get_first_digits(long num);
int get_digits(long num);

//Main method of the program that calls all other methods
int main(void)
{
    long number = get_credit_number();
    long copy = number;
    int sumDigits = get_digit_sum(number);
    int count = 1;
    int checkSum = 0;

    //Looping through the digits and adding every second digit to the checkSum afterwards
    do
    {
        //Validates if digits is every second one
        if (count % 2 == 0)
        {
            int r = 2 * (number % 10);

            if (get_length(r) > 1)
            {
                do 
                {
                    checkSum += r % 10;
                    r = r / 10;
                }
                while (r > 0);
            }
            else
            {
                checkSum += r;
            }

        }
        number = number / 10;
        count ++;
    }
    while (number > 0);

    //Adding the sum of the digits that were not multiplied by 2 to the checkSum
    checkSum += sumDigits;

    //Validating if the last digit of the check sum is 0
    checkSum % 10 != 0 ? printf("INVALID\n") : get_card_type(copy);

}
//Method that prompts the user for a credit card number
long get_credit_number(void)
{
    long number;

    //Prompting the user until they input a positive number
    do
    {
        number = get_long("Number: ");
    }

    while (number < 0);

    return number;
}

//Method that returns the length of each product
int get_length(long n)
{
    int count = 0;

    do
    {
        n = n / 10;
        count++;
    }
    while (n > 0);

    return count;
}

//Method that returns the sum of each digit not multiplied by 2
int get_digit_sum(long num)
{
    int count = 1;
    int sum = 0;

    //Looping through digits and adding every second digit to the sum variable
    do
    {
        if (count % 2 != 0)
        {
            sum += num % 10;
        }
        num = num / 10;
        count ++;
    }
    while (num > 0);

    return sum;
}
//Method that prints the type of credit card
void get_card_type(long n)
{
    int count = 0;
    long copy = n;
    int digit;
    string type = "";


    do
    {
        count++;
        digit = n % 10;
        n = n / 10;
    }
    while (n > 0);

    //Verifying if card is of type AMEX
    if (count == 15 && (get_digits(copy) == 34 || get_digits(copy) == 37))
    {
        type = "AMEX";
    }
    //Verifying if card is of type VISA
    else if (count == 13 && get_first_digits(copy) == 4)
    {
        type = "VISA";
    }
    //Verifying if card is of type VISA or MASTERCARD
    else if (count == 16)
    {
        if (get_first_digits(copy) == 4)
        {
            type = "VISA";
        }
        //Verifying if card is of type MASTERCARD
        else if (get_digits(copy) == 51 || get_digits(copy) == 52 || get_digits(copy) == 53 || get_digits(copy) == 54
                 || get_digits(copy) == 55)
        {
            type = "MASTERCARD";
        }
        else
        {
            type = "INVALID";
        }
    }
    else
    {
        type = "INVALID";
    }

    printf("%s\n", type);
}

//Method that returns the first two digits of the credit card number
int get_digits(long num)
{
    do
    {
        num /= 10;
    }
    while (num >= 100);
    return num;
}

//Method that returns the first digit of the credit card number
int get_first_digits(long num)
{
    do
    {
        num /= 10;
    }
    while (num >= 10);
    return num;
}